File: stock-exchange-api/StockExchange/Services/PortfolioService.cs
using StockExchange.Models;

namespace StockExchange.Services;

public class PortfolioService
{
    private decimal _cashBalance = 100000m;
    private readonly Dictionary<string, (int Quantity, decimal AverageCost)> _holdings = new();
    private readonly List<Order> _orders = new();
    private readonly List<TransactionEntry> _transactions = new();
    private readonly StockDataService _stockService;
    private readonly object _lock = new();

    public PortfolioService(StockDataService stockService)
    {
        _stockService = stockService;
        SeedPortfolio();
    }

    private void SeedPortfolio()
    {
        // start with an initial deposit transaction so transaction history explains starting cash
        var startingCash = _cashBalance;
        _transactions.Add(new TransactionEntry
        {
            Type = "DEPOSIT",
            Amount = startingCash,
            BalanceAfter = _cashBalance,
            Note = "Initial funding",
            CreatedAt = DateTime.UtcNow.AddDays(-90)
        });

        var seedTrades = new[] {
            ("AAPL", 10, 150.0m), ("MSFT", 5, 380.0m), ("NVDA", 3, 600.0m),
            ("GOOGL", 8, 140.0m), ("TSLA", 6, 200.0m)
        };
        foreach (var (symbol, qty, cost) in seedTrades)
        {
            _holdings[symbol] = (qty, cost);
            _cashBalance -= qty * cost;
            _orders.Add(new Order
            {
                Symbol = symbol,
                Name = _stockService.GetStock(symbol)?.Name ?? symbol,
                Type = "BUY",
                Quantity = qty,
                Price = cost,
                ExecutedAt = DateTime.UtcNow.AddDays(-_orders.Count * 7 - 30)
            });
            // record the cash outflow as a transaction entry
            _transactions.Add(new TransactionEntry
            {
                Type = "WITHDRAWAL",
                Amount = qty * cost,
                BalanceAfter = _cashBalance,
                Note = $"Seed buy {qty} {symbol}",
                CreatedAt = DateTime.UtcNow.AddDays(-_orders.Count * 7 - 30)
            });
        }
    }

    public Portfolio GetPortfolio()
    {
        lock (_lock)
        {
            var holdings = _holdings.Select(kvp =>
            {
                var stock = _stockService.GetStock(kvp.Key);
                return new Holding
                {
                    Symbol = kvp.Key,
                    Name = stock?.Name ?? kvp.Key,
                    Quantity = kvp.Value.Quantity,
                    AverageCost = kvp.Value.AverageCost,
                    CurrentPrice = stock?.Price ?? 0
                };
            }).ToList();

            return new Portfolio
            {
                CashBalance = _cashBalance,
                Holdings = holdings
            };
        }
    }

    public (bool Success, string Message, Order? Order) ExecuteTrade(TradeRequest request)
    {
        lock (_lock)
        {
            var stock = _stockService.GetStock(request.Symbol);
            if (stock == null) return (false, "Stock not found", null);
            if (request.Quantity <= 0) return (false, "Invalid quantity", null);

            var total = stock.Price * request.Quantity;

            if (request.Type.ToUpper() == "BUY")
            {
                if (total > _cashBalance) return (false, "Insufficient funds", null);
                _cashBalance -= total;

                if (_holdings.TryGetValue(request.Symbol, out var existing))
                {
                    var newQty = existing.Quantity + request.Quantity;
                    var newAvg = ((existing.AverageCost * existing.Quantity) + (stock.Price * request.Quantity)) / newQty;
                    _holdings[request.Symbol] = (newQty, Math.Round(newAvg, 4));
                }
                else
                {
                    _holdings[request.Symbol] = (request.Quantity, stock.Price);
                }
            }
            else if (request.Type.ToUpper() == "SELL")
            {
                if (!_holdings.TryGetValue(request.Symbol, out var holding)) return (false, "No position to sell", null);
                if (holding.Quantity < request.Quantity) return (false, "Insufficient shares", null);

                _cashBalance += total;
                var remaining = holding.Quantity - request.Quantity;
                if (remaining == 0) _holdings.Remove(request.Symbol);
                else _holdings[request.Symbol] = (remaining, holding.AverageCost);
            }
            else
            {
                return (false, "Invalid order type", null);
            }

            var order = new Order
            {
                Symbol = request.Symbol,
                Name = stock.Name,
                Type = request.Type.ToUpper(),
                Quantity = request.Quantity,
                Price = stock.Price
            };
            _orders.Add(order);
            // record cash movement for the trade
            _transactions.Add(new TransactionEntry
            {
                Type = request.Type.ToUpper() == "BUY" ? "WITHDRAWAL" : "DEPOSIT",
                Amount = total,
                BalanceAfter = _cashBalance,
                Note = $"{request.Type.ToUpper()} {request.Quantity} {request.Symbol}",
                CreatedAt = DateTime.UtcNow
            });
            return (true, "Order executed", order);
        }
    }

    public (bool Success, string Message, TransactionEntry? Transaction) Deposit(decimal amount)
    {
        lock (_lock)
        {
            if (amount <= 0) return (false, "Deposit amount must be greater than zero", null);

            _cashBalance += amount;
            var transaction = new TransactionEntry
            {
                Type = "DEPOSIT",
                Amount = amount,
                BalanceAfter = _cashBalance,
                Note = "Cash deposit"
            };
            _transactions.Add(transaction);
            return (true, "Deposit completed", transaction);
        }
    }

    public (bool Success, string Message, TransactionEntry? Transaction) Withdraw(decimal amount)
    {
        lock (_lock)
        {
            if (amount <= 0) return (false, "Withdrawal amount must be greater than zero", null);
            if (amount > _cashBalance) return (false, "Insufficient cash balance", null);

            _cashBalance -= amount;
            var transaction = new TransactionEntry
            {
                Type = "WITHDRAWAL",
                Amount = amount,
                BalanceAfter = _cashBalance,
                Note = "Cash withdrawal"
            };
            _transactions.Add(transaction);
            return (true, "Withdrawal completed", transaction);
        }
    }

    public IEnumerable<Order> GetOrders() 
    {
        lock (_lock)
            return _orders.OrderByDescending(o => o.ExecutedAt).ToList();
    }

    public IEnumerable<TransactionEntry> GetTransactions()
    {
        lock (_lock)
            return _transactions.OrderByDescending(t => t.CreatedAt).ToList();
    }
}

File: stock-exchange/src/app/services/portfolio.service.ts
import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable, Subject, combineLatest, map, of, throwError } from 'rxjs';
import { MoneyRequest, Portfolio, Holding, Order, TradeRequest, Transaction } from '../models/portfolio.model';
import { Stock } from '../models/stock.model';
import { StockService } from './stock.service';

const STORAGE_KEY = 'dragonfi:portfolio:v1';
const STARTING_CASH = 25000;

interface PersistedHolding {
  symbol: string;
  quantity: number;
  totalCost: number; // running sum of cost basis
}

interface PersistedState {
  cashBalance: number;
  holdings: PersistedHolding[];
  orders: Order[];
  transactions: Transaction[];
}

@Injectable({ providedIn: 'root' })
export class PortfolioService {
  private state$ = new BehaviorSubject<PersistedState>(this.empty());
  private tradeExecuted$ = new Subject<void>();

  constructor(private stockService: StockService) {
    this.state$.next(this.loadOrSeed());
  }

  // ─────────────────────────────────────────────────────────────────────────
  // Public API (compatible with previous HTTP-based service)
  // ─────────────────────────────────────────────────────────────────────────

  getPortfolio(): Observable<Portfolio> {
    return of(this.compute(this.state$.value, this.stockService.snapshot));
  }

  getLivePortfolio(): Observable<Portfolio> {
    return combineLatest([this.state$, this.stockService.live$]).pipe(
      map(([state, stocks]) => this.compute(state, stocks))
    );
  }

  executeTrade(request: TradeRequest): Observable<Order> {
    const stock = this.stockService.snapshot.find(s => s.symbol === request.symbol);
    if (!stock) return throwError(() => ({ error: { message: `Unknown symbol ${request.symbol}` } }));
    if (request.quantity <= 0) return throwError(() => ({ error: { message: 'Quantity must be positive' } }));

    const state = { ...this.state$.value };
    const price = stock.price;
    const total = round(price * request.quantity);

    if (request.type === 'BUY') {
      if (state.cashBalance < total) {
        return throwError(() => ({ error: { message: `Insufficient cash. Need ${fmt(total)}, have ${fmt(state.cashBalance)}.` } }));
      }
      state.cashBalance = round(state.cashBalance - total);
      const existing = state.holdings.find(h => h.symbol === stock.symbol);
      if (existing) {
        existing.quantity += request.quantity;
        existing.totalCost = round(existing.totalCost + total);
      } else {
        state.holdings = [...state.holdings, { symbol: stock.symbol, quantity: request.quantity, totalCost: total }];
      }
    } else {
      const existing = state.holdings.find(h => h.symbol === stock.symbol);
      if (!existing || existing.quantity < request.quantity) {
        return throwError(() => ({ error: { message: `Insufficient shares of ${stock.symbol}.` } }));
      }
      const costBasisPerShare = existing.totalCost / existing.quantity;
      existing.quantity -= request.quantity;
      existing.totalCost = round(existing.totalCost - costBasisPerShare * request.quantity);
      if (existing.quantity === 0) {
        state.holdings = state.holdings.filter(h => h.symbol !== stock.symbol);
      }
      state.cashBalance = round(state.cashBalance + total);
    }

    const order: Order = {
      id: makeId('ord'),
      symbol: stock.symbol,
      name: stock.name,
      type: request.type,
      quantity: request.quantity,
      price,
      total,
      executedAt: new Date().toISOString(),
      status: 'EXECUTED',
    };
    state.orders = [order, ...state.orders];

    this.commit(state);
    this.tradeExecuted$.next();
    return of(order);
  }

  getOrders(): Observable<Order[]> {
    return of(this.state$.value.orders);
  }

  getTransactions(): Observable<Transaction[]> {
    return of(this.state$.value.transactions);
  }

  deposit(amount: number): Observable<Transaction> {
    if (amount <= 0) return throwError(() => ({ error: { message: 'Amount must be positive' } }));
    const state = { ...this.state$.value };
    state.cashBalance = round(state.cashBalance + amount);
    const tx: Transaction = {
      id: makeId('tx'),
      type: 'DEPOSIT',
      amount: round(amount),
      balanceAfter: state.cashBalance,
      note: 'Deposit',
      createdAt: new Date().toISOString(),
      status: 'COMPLETED',
    };
    state.transactions = [tx, ...state.transactions];
    this.commit(state);
    return of(tx);
  }

  withdraw(amount: number): Observable<Transaction> {
    if (amount <= 0) return throwError(() => ({ error: { message: 'Amount must be positive' } }));
    const state = { ...this.state$.value };
    if (state.cashBalance < amount) {
      return throwError(() => ({ error: { message: `Insufficient cash. Available ${fmt(state.cashBalance)}.` } }));
    }
    state.cashBalance = round(state.cashBalance - amount);
    const tx: Transaction = {
      id: makeId('tx'),
      type: 'WITHDRAWAL',
      amount: round(amount),
      balanceAfter: state.cashBalance,
      note: 'Withdrawal',
      createdAt: new Date().toISOString(),
      status: 'COMPLETED',
    };
    state.transactions = [tx, ...state.transactions];
    this.commit(state);
    return of(tx);
  }

  notifyTradeExecuted() {
    this.tradeExecuted$.next();
  }

  get tradeExecuted(): Observable<void> {
    return this.tradeExecuted$.asObservable();
  }

  // ─────────────────────────────────────────────────────────────────────────
  // Internals
  // ─────────────────────────────────────────────────────────────────────────

  private compute(state: PersistedState, stocks: Stock[]): Portfolio {
    const priceMap = new Map(stocks.map(s => [s.symbol, s]));
    const holdings: Holding[] = state.holdings.map(h => {
      const s = priceMap.get(h.symbol);
      const currentPrice = s?.price ?? 0;
      const averageCost = h.quantity ? h.totalCost / h.quantity : 0;
      const currentValue = round(currentPrice * h.quantity);
      const gainLoss = round(currentValue - h.totalCost);
      const gainLossPercent = h.totalCost ? round((gainLoss / h.totalCost) * 100, 2) : 0;
      return {
        symbol: h.symbol,
        name: s?.name ?? h.symbol,
        quantity: h.quantity,
        averageCost: round(averageCost),
        currentPrice,
        totalCost: round(h.totalCost),
        currentValue,
        gainLoss,
        gainLossPercent,
      };
    });

    const totalInvested = round(holdings.reduce((sum, h) => sum + h.totalCost, 0));
    const totalMarketValue = round(holdings.reduce((sum, h) => sum + h.currentValue, 0));
    const totalValue = round(state.cashBalance + totalMarketValue);
    const totalGainLoss = round(totalMarketValue - totalInvested);
    const totalGainLossPercent = totalInvested ? round((totalGainLoss / totalInvested) * 100, 2) : 0;

    return {
      cashBalance: round(state.cashBalance),
      holdings,
      totalInvested,
      totalValue,
      totalGainLoss,
      totalGainLossPercent,
    };
  }

  private commit(state: PersistedState) {
    this.persist(state);
    this.state$.next(state);
  }

  private persist(state: PersistedState) {
    try { localStorage.setItem(STORAGE_KEY, JSON.stringify(state)); } catch {}
  }

  private loadOrSeed(): PersistedState {
    try {
      const raw = localStorage.getItem(STORAGE_KEY);
      if (raw) {
        const parsed = JSON.parse(raw) as PersistedState;
        if (parsed && typeof parsed.cashBalance === 'number') return parsed;
      }
    } catch {}
    const fresh = this.seed();
    this.persist(fresh);
    return fresh;
  }

  private empty(): PersistedState {
    return { cashBalance: STARTING_CASH, holdings: [], orders: [], transactions: [] };
  }

  private seed(): PersistedState {
    const now = new Date();
    const stocks = this.stockService.snapshot;
    const findPrice = (sym: string) => stocks.find(s => s.symbol === sym)?.price ?? 100;

    // Starter holdings — bought "yesterday" at slightly off prices so there's
    // some P&L for the user to see on first load.
    const seedTrades: Array<{ symbol: string; qty: number; pricePct: number }> = [
      { symbol: 'AAPL', qty: 12, pricePct: 0.97 },
      { symbol: 'MSFT', qty: 5,  pricePct: 1.02 },
      { symbol: 'NVDA', qty: 18, pricePct: 0.92 },
      { symbol: 'TSLA', qty: 4,  pricePct: 1.06 },
    ];

    let cash = STARTING_CASH + 25000; // start with more so we can buy and still have cash
    const holdings: PersistedHolding[] = [];
    const orders: Order[] = [];
    const transactions: Transaction[] = [];

    // initial deposit
    const seedDeposit: Transaction = {
      id: makeId('tx'),
      type: 'DEPOSIT',
      amount: cash,
      balanceAfter: cash,
      note: 'Initial funding',
      createdAt: new Date(now.getTime() - 3 * 24 * 60 * 60 * 1000).toISOString(),
      status: 'COMPLETED',
    };
    transactions.push(seedDeposit);

    seedTrades.forEach((t, i) => {
      const stock = stocks.find(s => s.symbol === t.symbol);
      if (!stock) return;
      const buyPrice = round(findPrice(t.symbol) * t.pricePct);
      const total = round(buyPrice * t.qty);
      if (cash < total) return;
      cash = round(cash - total);
      holdings.push({ symbol: t.symbol, quantity: t.qty, totalCost: total });
      orders.push({
        id: makeId('ord'),
        symbol: t.symbol,
        name: stock.name,
        type: 'BUY',
        quantity: t.qty,
        price: buyPrice,
        total,
        executedAt: new Date(now.getTime() - (2 - i * 0.1) * 24 * 60 * 60 * 1000).toISOString(),
        status: 'EXECUTED',
      });
    });

    return {
      cashBalance: cash,
      holdings,
      orders,
      transactions,
    };
  }
}

let _idCounter = 0;
function makeId(prefix: string): string {
  _idCounter += 1;
  return `${prefix}_${Date.now().toString(36)}_${_idCounter.toString(36)}`;
}

function round(n: number, p = 2): number {
  const f = Math.pow(10, p);
  return Math.round(n * f) / f;
}

function fmt(n: number): string {
  return `$${n.toFixed(2)}`;
}
